ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,864 @@
+      IndexSegment indexSegmentToCopy = srcIndex.getIndexSegments().get(indexSegmentStartOffset);
+      if (needsCopying(indexSegmentToCopy.getEndOffset()) && !copyIndexSegmentData(logSegmentToCopy, indexSegmentToCopy,
+          duplicateSearchSpan, startOffsetOfLastIndexSegmentForDelete)) {
+        // there is a shutdown in progress or there was no space to copy all entries.
+        allCopied = false;
+        break;
+      }
+    }
+    return allCopied;
+  }
+
+  /**
+   * Copies data in the provided {@code indexSegmentToCopy} into the target log (swap spaces).
+   * @param logSegmentToCopy the {@link LogSegment} to copy from.
+   * @param indexSegmentToCopy the {@link IndexSegment} that contains the entries that need to be copied.
+   * @param duplicateSearchSpan the {@link FileSpan} in which to search for duplicates.
+   * @param startOffsetOfLastIndexSegmentForDelete  the start {@link Offset} of the index segment up until which delete
+   *                                                records are considered applicable.
+   * @return {@code true} if all the records in the index segment were copied. {@code false} if some records were not
+   * copied either because there was no more capacity or because a shutdown was initiated.
+   * @throws IOException if there were I/O errors during copying.
+   * @throws StoreException if there are any problems reading or writing to store components.
+   */
+  private boolean copyIndexSegmentData(LogSegment logSegmentToCopy, IndexSegment indexSegmentToCopy,
+      FileSpan duplicateSearchSpan, Offset startOffsetOfLastIndexSegmentForDelete) throws IOException, StoreException {
+    List<IndexEntry> allIndexEntries = new ArrayList<>();
+    // call into diskIOScheduler to make sure we can proceed (assuming it won't be 0).
+    diskIOScheduler.getSlice(INDEX_SEGMENT_READ_JOB_NAME, INDEX_SEGMENT_READ_JOB_NAME, 1);
+    // get all entries
+    indexSegmentToCopy.getIndexEntriesSince(null, new FindEntriesCondition(Long.MAX_VALUE), allIndexEntries,
+        new AtomicLong(0));
+
+    // save a token for restart (the key gets ignored but is required to be non null for construction)
+    StoreFindToken safeToken =
+        new StoreFindToken(allIndexEntries.get(0).getKey(), indexSegmentToCopy.getStartOffset(), sessionId,
+            incarnationId);
+    compactionLog.setSafeToken(safeToken);
+
+    boolean checkAlreadyCopied =
+        recoveryStartToken != null && recoveryStartToken.getOffset().equals(indexSegmentToCopy.getStartOffset());
+
+    // filter deleted/expired entries and get index entries for all the PUT and DELETE records
+    List<IndexEntry> indexEntriesToCopy =
+        getIndexEntriesToCopy(allIndexEntries, duplicateSearchSpan, indexSegmentToCopy.getStartOffset(),
+            startOffsetOfLastIndexSegmentForDelete, checkAlreadyCopied);
+
+    // Copy these over
+    boolean copiedAll = copyRecords(logSegmentToCopy, indexEntriesToCopy, indexSegmentToCopy.getLastModifiedTimeSecs());
```

Comment:
I am bit uncomfortable as our checkpointing works only on indexSegment boundaries. We should definitely work (even if not in this patch, but in future) on improving it. For instance, once in 5K(config) elements, we should persist the targetIndex and checkpoint the token. So that during recovery we can move the token directly to where we left and start from there instead of starting from index startOffset. 
Note: I do know that even though we start from startOffset for now, we don't do double copying at any time as do check if its present in targetIndex via alreadyExistsInTgt. But even then, trying to see if we can optimize that